sklearn/linear_model/omp.py

```diff
@@ -806,6 +806,7 @@ def fit(self, X, y):
             returns an instance of self.
         """
         X, y = check_X_y(X, y)
+        X = as_float_array(X, copy=False)
```

Comment:
Actually we should also pass `force_all_finite=False` as this check is already done in `check_X_y` and it's not free.

That would be another motivation to have a richer `check_X_y` / `check_array`. We could have for instance:

```python
X, y = check_X_y(X, y, dtype=[np.float64, np.float32])
```

If `dtype` is a `list` or `tuple` and `X.dtype` is not in that list, the data would be cast to `dtype[0]`. This would be similar to the behavior of `accept_sparse`.